0010-R-components-healthcheck.md

```diff
@@ -0,0 +1,477 @@
+We should definitely have some Work Items to ensure that most Components start implementing Ping, but that doesn't mean that we can't go ahead with Components health check for now, we just need to ensure that we are providing a proper info about those components explicitly to user which are in his system but could not be verified via the health check.
+
+
+## API Design
+
+Most of the Approaches underneath work with a query parameter, in addition to `healthz` endpoint. Instead of, an additional endpoint.
+
+Following are the possible responses for `healthz` API:
+
+| HTTP | Response Codes | 
+| -------- | -------- | 
+| 204     | dapr is healthy     | 
+| 500     | dapr is not healthy     | 
+
+Hence, if dapr is healthy, then the response code changes, as per the enlisted case below.
+If dapr is NOT healthy, then the compoenents Health check should anyways NOT be checked.
+
+
+### Approach 1:
+List only failed or tose componnets for which Ping is not implemented. Create separate sections for "Failed" and "Ping Not Implemented".
+
+HTTP Endpoint: http://localhost:3500/v1.0/healthz?components=true
+GRPC Endpoint: GetComponentHealthAlpha1 (It will be separate endpoint for grpc)
+
+**Case 1:** When ALL components in system implement Ping and are Healthy
+
+Result:
+
+Response code: 204 OK No Content
+GRPC Response code (if implemented): 0 OK
+
+**Case 2:** When ALL components in system implement Ping BUT some components have failed Ping check: Here we report only failed components and not passed components, we don't treat this API as a way to list components.
+
+HTTP Response code: 500 Internal Server Error
+GRPC Response code (if implemented): 2 UNKNOWN
+
+```
+{
+    "componentsHealth": 
+        "failedComponents": [
+            {
+                "componentName": "txnstore",
+                "type": "state",
+                "status": "NOT_OK",
+                "message": "redis store: error connecting to redis at localhost:6379: dial tcp 127.0.0.1:6379: connect: connection refused"
+            },
+            {
+                "componentName": "smsbinding",
+                "type": "bindings",
+                "status": "NOT OK",
+                "message": "redis binding: error connecting to redis at localhost:6379: dial tcp 127.0.0.1:6379: connect: connection refused"
+            }
+        ]
+}
+```
+**Case 3:** When SOME components in system implement Ping AND some components DON'T implement Ping, AND some components have failed Ping check as well: Here we report failed components and those components as well for which Ping is not implemented, we don't treat this API as a way to list components.
```

Comment:
Comment from Alessandro (@ItalyPaleAle ): I am not sure we need a case for this. As per my other comment, IMHO it's totally fine for a component to NOT have Ping implemented. I think the healthcheck should invoke Ping when available, and otherwise just skip the component

Ref: https://hackmd.io/MaNpUYRyQqe-0eqCsSUiIg

Reply:
But, won't it give False Positives then? How as a user am I supposed to know that what all components implement Ping and  that only those components are covered by this Health Check?